Привет, меня зовут Артем Хуршудов. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

### <font color='orange'>Общее впечатление</font>
* Большое спасибо за проделанную работу. Видно, что приложено много усилий.
* Отлично, что правильно использованы train/valid/test наборы.
* Выводы и рассуждения получились содержательными, их было интересно читать.
* Над этим проектом нужно будет еще немного поработать. Однако, изменения не должны занять много времени.
* С радостью отвечу на твои вопросы, если они есть. Лучше всего их собрать в следующей ячейке. Жду новую версию проекта.

### <font color='orange'>Общее впечатление (ревью 2)</font>
* Для удобства все новые комментарии обозначены фразой "ревью 2".
* Удачи в доработке!

### <font color='orange'>Общее впечатление (ревью 3)</font>
* Новый комментарий обозначен фразой "ревью 3".

### <font color='orange'>Общее впечатление (ревью 4)</font>
* После исправлений проект улучшился и теперь он может быть зачтен.
* Удачи в дальнейшем обучении и следующих работах!

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

## Откройте и изучите файл

In [31]:
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.dummy import DummyClassifier
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import RandomForestClassifier
import pandas as pd
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
df = pd.read_csv('/datasets/users_behavior.csv')

<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что все импорты собраны в первой ячейке ноутбука! Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
</div>

In [32]:
df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

Посмотрим тип данных

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
calls       3214 non-null float64
minutes     3214 non-null float64
messages    3214 non-null float64
mb_used     3214 non-null float64
is_ultra    3214 non-null int64
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


Изменим тип поля messages на int, потому что дробных сообщений быть не может

In [34]:
df['messages']=df['messages'].astype(int)

In [35]:
df['messages']

0       83
1       56
2       86
3       81
4        1
        ..
3209    20
3210     0
3211    70
3212    90
3213     6
Name: messages, Length: 3214, dtype: int64

Проверим дубликаты

In [36]:
df.duplicated().sum()

0

In [37]:
df.columns

Index(['calls', 'minutes', 'messages', 'mb_used', 'is_ultra'], dtype='object')

Проверим значения содержащиеся в столбцах

In [38]:
df.describe()

calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

ВЫВОД:
1. Пропущенных значений нет
2. Типы столбцов в порядке
3. Дубликатов нет

<div class="alert alert-block alert-success">
<b>Успех:</b> Данные загружены корректно, первичный осмотр проведен.
</div>

## Разбейте данные на выборки

Разобьем выборки на  60:20:20.


In [39]:
#train, test = train_test_split(df, test_size=0.4, random_state=12345)


In [40]:
train_df, othr_df = train_test_split(df, test_size=0.40, random_state=12345)
valid_df, test_df = train_test_split(vnt_df, test_size=0.50, random_state=12345)

In [62]:
print('Размер тренировочной выборки', train_df.shape[0])
print('Размер валидационной выборки', valid_df.shape[0])
print('Размер тестовой выборки', test_df.shape[0])

Размер тренировочной выборки 1928
Размер валидационной выборки 643
Размер тестовой выборки 643


In [63]:
train_df_features = train_df.drop(['is_ultra'], axis=1)
train_df_target = train_df['is_ultra']
valid_df_features = valid_df.drop(['is_ultra'], axis=1)
valid_df_target = valid_df['is_ultra']
test_df_features = test_df.drop(['is_ultra'], axis=1)
test_df_target = test_df['is_ultra']

In [43]:
#features = df.drop(['is_ultra'], axis=1)
#target = df['is_ultra']


#Необходимо создать тестовую выборку и валидационную 20% на 20 %

In [44]:
#features_val, features_test, target_val, target_test = train_test_split(features, target, test_size=0.2, random_state=12345)
#features_train, features_test, target_train, targe_test = train_test_split(features,  target, test_size=0.8)
#features_val, features_train,  target_val, target_train, = train_test_split(features,target, test_size=0.8, random_state=12345)
#x_test, x_val, y_test, y_val = train_test_split(x_test, y_test, test_size=0.5)

In [46]:
#features_train = train.drop('is_ultra', axis=1)
#target_train = train['is_ultra']


In [47]:
#features_train.info()


In [48]:
#features_test.info()


In [49]:
#features_val.info()


features_train, target_train - обучающая выборка

features_val, target_val - валидационная выборка

features_test, target_test - тестовая выборка


<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Пожалуйста, напечатай размеры каждой из выборок и убедись, что ты сделал все верно.
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Проверил размеры воборки, все совпало, всего у нас 3214, в "тренировочной выборке" 1928, что соответствует 60 % и  1607,что 40 %. 
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка (ревью 2):</b> 1928 + 1607 = 3535, а не 3214. А размер валидационной? Все выборки должны быть независимыми, то есть не иметь общего объекта. Ошибка заключается в коде, который создает валидационную выборку.
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Артем, спасибо! За исправления, я вчера, когда отправил, завис с этим моментом, только после отправки начал думаь,ч то что-то не так,  и как выяснилось не зря думал...Мои коллеги по группе вообще для выделения использовали дополнительный импорт файлов, я хотел использовать именно все данные, которые дали в теоретическом курсе, проверь пожалуйста, сейчас вроде так.
</div>

<div class="alert alert-block alert-danger">

<b>Ошибка (ревью 3):</b> Не совсем. Давай сделаем так: 
1. Изначальный датафрейм разделим на features и target.
2. Полученные наборы разобьем на train и other в соотношении 60:40.
3. other разобьем на valid, test в соотношении 50:50.
4. В итоге получится 60:20:20.
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Артем, спасибо! Надеюсь сейчас верно тебя понял! Я чуть переименовал переменный, т.к. уже начал запутываться...
</div>

## Исследуйте модели

Дерево решений

In [50]:
best_accuracy = 0
best_depth = 0
for depth in range(1,101):
    model = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model.fit(train_df_features, train_df_target)
    accuracy = model.score(valid_df_features, valid_df_target)
    if accuracy > best_accuracy:
        # print("Глубина дерева:", depth,"    Качество:" ,accuracy)
        best_accuracy = accuracy
        best_depth = depth
print("Глубина дерева:", best_depth, "    Качество:", best_accuracy)

Глубина дерева: 3     Качество: 0.7853810264385692


<div class="alert alert-block alert-success">
<b>Успех:</b> ПОдбор параметров реализован правильно.
</div>

Логическая регрессия

In [51]:
model = LogisticRegression()
model.fit(train_df_features, train_df_target)
accuracy = model.score(valid_df_features, valid_df_target)
print("Качество:", accuracy)

Качество: 0.7589424572317263


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Линейная регрессия

In [52]:
#model = LinearRegression()  # инициализируйте модель LinearRegression
#model.fit(features_train, target_train) # обучите модель на тренировочной выборке
#accuracy = model.score(features_val, target_val)
#print("Качество:", accuracy)

Какой-то не правдободобный вывод, думаю линейную регрессию брать нельзя...

<div class="alert alert-block alert-warning">
<b>Изменения:</b>Где можно почитать информацию, в каком случае используется линенйая регрессия.
</div>

<div class="alert alert-block alert-info">

<b>Совет (ревью 2): </b> Линейная регрессия – алгоритм для решения задачи регрессии. В данном случае ее применении не имеет особого смысла, так как у нас задача классификации. Более того, `.score` возвращает не accuracy в данном случае.
</div>

Случайный лес

In [53]:
best_accuracy = 0
best_est = 0
best_depth = 0
for est in range(10, 60, 10):
    for depth in range (1, 10):
        model = RandomForestClassifier(max_depth=depth, n_estimators=est, random_state=12345)
        model.fit(train_df_features, train_df_target)
        accuracy = model.score(valid_df_features, valid_df_target)
        if accuracy > best_accuracy:
            best_accuracy = accuracy
            best_depth = depth
            best_est = est
print("Глубина дерева:", best_depth, "    Количество деревьев:", best_est,  "  Качество:", best_accuracy)

Глубина дерева: 8     Количество деревьев: 40   Качество: 0.8087091757387247


<div class="alert alert-block alert-success">
<b>Успех:</b> Радует, что в данном случае ты выбираешь сразу несколько параметров.
</div>

## Проверьте модель на тестовой выборке

In [54]:
model = DecisionTreeClassifier(max_depth=11, random_state=12345)
model.fit(train_df_features, train_df_target)
accuracy = model.score(test_df_features, test_df_target)
print("Качество:", accuracy)

Качество: 0.7744945567651633


In [55]:
model = LogisticRegression()
model.fit(train_df_features, train_df_target)
accuracy = model.score(test_df_features, test_df_target)
print("Качество:", accuracy)

Качество: 0.7402799377916018


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [56]:
model = RandomForestClassifier(max_depth=7, n_estimators=40,random_state=12345)
model.fit(train_df_features, train_df_target)
accuracy = model.score(test_df_features, test_df_target)
print("Качество:", accuracy)

Качество: 0.80248833592535


ВЫВОД:

Наилучшее качество (0.816) достигается при использовании дерево решений с гулбиной

<div class="alert alert-block alert-warning">
<b>Изменения:</b>Получается вывод поменялся на лес решений с качеством 0,802
</div>

<div class="alert alert-block alert-success">
<b>Успех:</b> Тестирование было сделано абслютно верно, молодец!
</div>

## (бонус) Проверьте модели на адекватность

In [57]:
dmb = DummyClassifier(strategy="stratified" ,random_state=12345)
dmb.fit(train_df_features, train_df_target)
acc = dmb.score(test_df_features, test_df_target)
print("Качество наивного прогноза :", acc)
print("Качество нашего прогноза :", accuracy)

Качество наивного прогноза : 0.536547433903577
Качество нашего прогноза : 0.80248833592535


In [58]:
dmb = DummyClassifier(strategy='most_frequent', random_state=12345)
dmb.fit(train_df_features, train_df_target)
acc = dmb.score(test_df_features, test_df_target)
print("Качество наивного прогноза :", acc)
print("Качество нашего прогноза :", accuracy)

Качество наивного прогноза : 0.6842923794712286
Качество нашего прогноза : 0.80248833592535


In [59]:
dmb = DummyClassifier(strategy='uniform', random_state=12345)
dmb.fit(train_df_features, train_df_target)
acc = dmb.score(test_df_features, test_df_target)
print("Качество наивного прогноза :", acc)
print("Качество нашего прогноза :", accuracy)

Качество наивного прогноза : 0.4821150855365474
Качество нашего прогноза : 0.80248833592535


Вывод:
    
    Наилучшее качество (0.816) достигается при использовании дерево решений с гулбиной 10.

Вывод:
    
<div class="alert alert-block alert-warning">
<b> Наилучшее качество (0.802) достигается при использовании леса решений с гулбиной 8 и кол-во деревьев 40:</b> ...
</div> 
  
 .

<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, именно это и есть проверка на адеватность.
</div>

Вывод:
    
<div class="alert alert-block alert-warning">
<b> Я правильно понял, изменение размеров выборки повлияло на обучение нашей модели.</b> ...
</div> 
  
 .

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75
